### ISSUE TITLE
Returning Observable through processor and how to move jobs to error vs completed.

### ISSUE THREAD

**Issue by camcramz** [taskforcesh/issues#151]
I have a free trial i am working with but my company has purchased a license for us to move forward with.  We are building an environment where worker's business logic are configurable and I want to be able to pass the job onward and have the configured business logic determine whether a job has succeeded or failed.  An observable seemed to fit the bill for this but how do i communicate back to the observable when a job should fail vs succeed?  How does the WorkerPro class tie into the observable.  Do i override the error and complete methods and move the job manually there or am i able to just call those and WorkerPro picks them up from there?

**Comment by manast** [comment]
@camcramz I am not sure I completely follow your use-case. Observables in the context of BullMQ processors are mainly to allow you to cancel job in a clean way (as promises do not support this), or in order to implement jobs that have different steps that could fail, and thus the Observable could store the last step so that a retry would continue from where it left (as I guess you have already read https://docs.bullmq.io/bullmq-pro/observables).

If you could expand a bit more on your use case, maybe with some pseudo-code I would be in a better position to help you.

**Comment by camcramz** [comment]
Thank you for the response!

I'm building a dynamic queuing mechanism that creates queues for implemented business logic.  I have an express server that I mutate during runtime to add additional API endpoints to it.  The endpoint queues incoming data and I want to have a worker that picks it up and then passes it off to the script that runs the business logic.  I don't have anyway to know the completion of the script back to the worker but do not want to automatically pick up the next job until the script was complete.  So i was hoping I could subscribe to the obserable and pass the subscriber and job onward to the script for it to notify back to the observer that the job was complete and the next job could be continued.

const processor = async (job) => {
      return new Observable(subscriber => {
        const msg = {};
        msg._job = job;
        msg._subscriber = subscriber;
        this.send(msg);
      })
    }
    
    This would be an example of the processor i am working with.

And then within the script I would be able to either perform _subscriber.complete() or _job.moveToCompleted (Or to failed).  This script will handle all exceptions thrown.

Does this help?

**Comment by camcramz** [comment]
Also after sleeping on this.  I think what i really want to understand is what WorkerPro is doing with the returned observable and how it interacts with it?

**Comment by manast** [comment]
> I don't have anyway to know the completion of the script back to the worker but do not want to automatically pick up the next job until the script was complete

I think you could just use groups for this, with max concurrency of 1, that means that only one job will ever be running for a given group. So the job processes, runs the script and when the script completes so does the job.



> I have an express server that I mutate during runtime to add additional API endpoints to it

Not sure what mutate means here, but if you mean add/remove endpoints in runtime I never heard something like that. Would't you be able to achieve a similar functionality just relying on a fix route and use different params for the different "dynamic endpoints" ?

**Comment by manast** [comment]
Here more info about the max concurrency for groups: https://docs.bullmq.io/bullmq-pro/groups/concurrency

**Comment by camcramz** [comment]
Yeah that could be a similar solution.  We have a webserver written in express and we can add and splice routes while the application is running.  The webserver then is configured to pass off the requests to business logic that can be written on the fly.  Since this is configurable there's no way to return any thing to the worker or await the completion in the worker.

Is there a way to test the value returned by subscriber.next() to then complete or fail the job.  The current problem i have with doing this is a lock mismatch when i try to call the moveToComplete or moveToFailed method.  I've tried to pass the token from the pulled job and pass that through but it still throws an exception.  So hopefully leveraging the subscriber and understanding how that ultimately dictates to the workerpro when a job is finished is what i'm trying to understand.  Since you mentioned that a job can be cancelled through this what would be the way to cancel a running job?  And could that be used to move it through the queue to finished or failed?

**Comment by manast** [comment]
I am sorry but I cannot comprehend your case, it sounds overly complicated, I do not think you need the observable functionality of BullMQ for this either.

**Comment by camcramz** [comment]
The observable seems to work for me...except for when deciding when jobs fail and succeed a new job isn't automatically picked up.  Here is my processor:
```ts
const processor = async (job) => {
  const observable = new Observable();
  const subscriber = observable.subscribe({
    next: (x) => {
      console.log(x);
    },
    complete: (x) => {
      job.moveToCompleted({message: 'success'}, job.token);
    },
    error: (x) => {
      job.moveToFailed({message: 'failed'}, job.token);
    }
  });

  const msg = {};
  msg._job = job;
  msg._subscriber = subscriber;
  this.send(msg);
  
  return observable;
}
```

And then i attach the subscriber to the message being sent to my business logic.  Upon completion or if any errors are encountered the business logic can through the subscriber call complete or error.  This works for my use case but now disconnects me from taking advantage of using the worker to accept new jobs.  

I implemented queueEvents and can watch for completions or failures and call next job.  But what would be the best way to pick back up jobs once the queue has been depleted and a new job comes in?  Do i have to get rid of passing the processor into the worker and take the full manual approach?

**Comment by manast** [comment]
When a processor returns an observable, it is in fact the complete callback that triggers moveToCompleted, likewise when the observable errors it calls moveToFailed, most likely you are getting an unhandled exception when both your code and the worker tries to moveToComplete at the same time.

**Comment by manast** [comment]
The call ```this.send(msg)```, is it asynchronous? I am a bit confused on why you need this call instead of just running the code inside the processor.

**Comment by manast** [comment]
Regarding using manual fetch for jobs, this solution will be not very robust and scale to more than one machine as you will not be able to guarantee your requirement of only running one job at a time. If you only run on one machine and you do not need a very robust solution you could just make sure you only have one worker running at any given moment, and use concurrency = 1.

**Comment by camcramz** [comment]
to implement our business logic we are writing a plugin for a flowgramming tool.  The tool has blocks of code tied together by connections and because this is just one piece to instantiate a flow it won't have the ability to understand when the full flow was complete.  The reason observables were so intriguing to me is for the same reason it probably was with canceling a job where promises wouldn't help.  The this.send is purely passing the msg object to the next block of code and get's no return.

We will have a scaled system and the requirement is one per worker at a time per system, or have it configurable to where that number could be increased to allow additional jobs through.  So scalability is required.

I will try just a default Observable again and see what happens when i just call subscriber.complete() or .error() later in the flow.

**Comment by manast** [comment]
You can use observables to implement cancellation, but I think you should use groups (maybe just 1 group if that is what you need) and activate the max concurrency per group to 1. That will guarantee only one job per group at a time and it will also scale if you need.

**Comment by camcramz** [comment]
Thank you for the feedback, sorry about the delay in getting back to you.  With using that configuration, in my use case, where jobs complete outside the processor how do i ensure the next job isn't picked up until it is indicated that the work being performed is actually complete?

**Comment by manast** [comment]
Not sure what you mean by "complete outside the processor", can you provide more context?
